### ISSUE TITLE
Random attribute with Distinct and wide range causes test to disappear

### ISSUE THREAD

**Issue by ploeh** [nunit/nunit#4308]
# Steps to reproduce

Create a new NUnit test project with this test class:

```csharp
using NUnit.Framework;

namespace Repro;

public class Tests
{
    [Test]
    public void Test1(
        [Random(int.MinValue / 2, int.MaxValue / 2 + 1, count: 10, Distinct = true)] int i)
    {
        Assert.Pass();
    }
}
```

Run it with `dotnet test`.

# Expected behaviour
The test runs and passes ten time.

# Actual behaviour
No tests run.

```
No test is available in [redacted]\Repro\Repro\bin\Debug\net7.0\Repro.dll.
Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate
and try again.
```

# More information
This seems to be related to the requested range of the random values. If you remove the `+ 1` the tests execute as expected:

```csharp
[Test]
public void Test1(
    [Random(int.MinValue / 2, int.MaxValue / 2, count: 10, Distinct = true)] int i)
{
    Assert.Pass();
}
```

Alternatively, if you leave out the `Distinct` requirement, the tests also run as expected:

```csharp
[Test]
public void Test1(
    [Random(int.MinValue / 2, int.MaxValue / 2 + 1, count: 10)] int i)
{
    Assert.Pass();
}
```

System information:
- Windows 11 Pro 22H2
- .NET 7

**Comment by mikkelbu** [comment]
Thanks for the report @ploeh.

The problem is in the following check (also exists for the other datatypes)

https://github.com/nunit/nunit/blob/30e1255feb2c7108e6b7b168126360ac3689d4bb/src/NUnitFramework/framework/Attributes/RandomAttribute.cs#L349-L352

when `max = int.MaxValue / 2` then the `max - min` is 2147483647, but for `max = int.MaxValue / 2 + 1` we wrap around and `max - min` becomes `-2147483648` resulting in `CanBeDistinct` returning false.

Ps. Enjoyed "Code That Fits in Your Head". It gave me a lot to think about.

**Comment by skewl84** [comment]
Just looking at this issue to see in terms of a solution. Think it'd be right to 
- Do the return statement within a "checked" block  as that would throw a "System.OverflowException" in the case of an Arithmetic operation overflow or
- Throw an exception in the case of the max value being less than the min value. 

Please see the suggestion below:

` protected override bool CanBeDistinct(int min, int max, int count) 
 { 
    if (max >= min)
    {
        checked
        {
            return count <= max - min;
        }
    }
    else {
	throw new Exception($"Max value ({max}) provided is less than the Min value ({min})");
    } 
 }  `
 
 This approach should work for all the other supported data types as well.

**Comment by stevenaw** [comment]
Thanks for your interest @skewl84 !

> Throw an exception in the case of the max value being less than the min value.

This is a good point. It looks like this may be some missing validation at the moment, but there's a few different places this could be tackled. Often in cases like this we've added a roslyn analyzer to the `NUnit.Analyzers` package to give development-time feedback on if invalid constants are passed into various NUnit APIs, but another option could just be to "swap" them internally to ensure that `max` is always larger. I'd be curious to hear @nunit/framework-team  thoughts on this.

For now, I'd recommend to keep this simple and assume that `max` will always be larger than `min` :)


> Do the return statement within a "checked" block as that would throw a "System.OverflowException" in the case of an Arithmetic operation overflow

This is a nice idea, but I'd suggest that it may also hide some valid cases. Generating a subset of random values between `-1073741824` and `1073741829` (`int.MinValue / 2` and `int.MaxValue / 2 + 1` respectively) is a big range but also one that someone may want or need supported. A unhandled exception inside of the `RandomAttribute` itself would likely cause the test method to silently not run - something that may be a surprise.

There's a few different ways to handle this, but the simplest may be to avoid overflowing by casting one of the two values to a larger datatype. For example, building off @mikkelbu 's link in the code above, casting an `int` to a `long` would prevent the overflow but still allow the calculation to complete correctly

```diff
protected override bool CanBeDistinct(int min, int max, int count) 
 { 
-     return count <= max - min; 
+     return count <= (long)max - min; 
 } 
```

This approach won't work for `long` and `ulong` as there _is_ no larger datatype in `netstandard`, but in that case an overflowed negative value - if it hadn't overflowed - would also almost certainly be larger than the `int count` it's being compared to. In that case `CanBeDistinct` can likely safely just return `true`.

**Comment by skewl84** [comment]
@stevenaw Sounds good to me, as this is in a manner an edge case scenario. Happy to look at the NUnit.Analyzers package if that helps and Thanks for breaking this down.

**Comment by stevenaw** [comment]
@skewl84 Happy to help!
To be clear: there is definitely still a bugfix needed in the framework itself here in this repo to avoid an overflow for a particularly large range. If you're interested in contributing then you're very welcome to pick up this issue.

To my knowledge neither the analyzers or framework support validating the min/max arguments to the Range attribute, though I'm suggesting we hear from more team members about "where" that fix should be done before anyone spends time fixing it.

**Comment by skewl84** [comment]
sure thing. @stevenaw

**Comment by mr-russ** [comment]
If the key is a simple solution, this changes the arithmetic so you don't get overflow unless you have min + count overflow, with will be a weird result as max must be less tnan that.  The first check here  ensures if min and coynt overflow, is can't be unique and since we have that protection the second portion is safe without overflow.
```
protected override bool CanBeDistinct(int min, int max, int count) 
 { 
-     return count <= max - min; 
+     return min + count > min && min + count <= max; 
 }
```
Thoughts?  I can make a pull request if this is a reasonable solution.

**Comment by stevenaw** [comment]
Thanks @mr-russ 
I'm not 100% certain, but I think @skewl84 may already be working on this. @skewl84 is that right? I can assign you the issue if you are for clarity

**Comment by skewl84** [comment]
@stevenaw no I haven't but I can definitely work on this. I was under the impression to make sure the approach is right and we can move ahead. So if that can be agreed by the other members that would be great.

**Comment by mr-russ** [comment]
My outside opinion is the fix is required, and the discussion about Analyzers seems separate and additional.  I think there is value in submitting the fix and test for that first.  If there is consensus about further value, add that later.

**Comment by stevenaw** [comment]
Sorry for the confusion @skewl84 
That's right,  I think the analyzers discussion and how to validate that "max is greater than min" is separate from the underlying overflow issue here in the framework and that there's still value in fixing the overflow issue first.
I can assign you to the issue now

**Comment by mikkelbu** [comment]
As far as I can tell @mr-russ suggestion in https://github.com/nunit/nunit/issues/4308#issuecomment-1465887889 is correct (note that we probably also need the same fix for some of the other types - long, ulong etc.).

I'll create a separate issue for the Analyzers for Min >= Max (I've the feeling - I've not tested it - that if we throw here, then it will not only make the decorated test disappear but also the other tests in the file. Furthermore, I don't think we can show the problem to the users as there is no "test" to associated this error to).

**Comment by skewl84** [comment]
@mikkelbu  yep that'd be the right approach. @stevenaw could you please assign this to @mr-russ in this case. Thanks

**Comment by stevenaw** [comment]
Sure.
@mr-russ are you still interested in contributing your fix for this?

**Comment by mr-russ** [comment]
@stevenaw I'll begin with a master branch, how does backpatching work?

**Comment by mr-russ** [comment]
After working on this code, we have https://github.com/nunit/nunit/blob/4326ba6add8197fc2da7a58f2e45e70fd6ac3a50/src/NUnitFramework/framework/Attributes/RandomAttribute.cs#L250 , which has guards for some of the Random checks.  Is that a good spot to add min < max and count > 0 checks?
